docs/Configuration-Settings.md
Page\: `[Gear Icon] -> Configure`

The settings menu, accessed by selecting the gear icon in the top-right, then the Configure link, is a general area for various system-wide configuration options.

## General Settings

Page\: `[Gear Icon] -> Configure -> General`

<table>
<thead>
<tr class="header">
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Language</td>
<td>Set the language that will be displayed in the web user interface.</td>
</tr>
<tr>
<td>Force HTTPS</td>
<td>Require web browsers to use SSL/HTTPS. Any request to <a href="http://">http://</a> will be redirected to <a href="https://">https://</a>.</td>
</tr>
<tr>
<td>Hide success alerts</td>
<td>Hide all success alert boxes that appear at the top of the page.</td>
</tr>
<tr>
<td>Hide info alerts</td>
<td>Hide all info alert boxes that appear at the top of the page.</td>
</tr>
<tr>
<td>Hide warning alerts</td>
<td>Hide all warning alert boxes that appear at the top of the page.</td>
</tr>
<tr>
<td>Opt-out of statistics</td>
<td>Turn off sending anonymous usage statistics. Please consider that this helps the development to leave on.</td>
</tr>
</tbody>
</table>

## Time Series Database Settings

Page\: `[Gear Icon] -> Configure -> General`

Measurements are stored in a time-series database. There are currently two options that can be used with Mycodo, InfluxDB 1.x and InfluxDB 2.x. InfluxDB 1.x works on both 32-bit and 64-bit operating systems, but 2.x only works on 64-bit operating systems. Therefore, if you are using a 32-bit operating system, you will need to use InfluxDB 1.x. During the Mycodo install, you can select to install influxDB 1.x, 2.x, or neither. If you don't install InfluxDB, you will need to specify the host and credentials to an alternate install for Mycodo to be able to store and query measurements.

If you are installing via Docker, you will need to change the hostname to "mycodo_influxdb" after the Mycodo install to be able to connect to the InfluxDB Docker container.

<table>
<thead>
<tr class="header">
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Database</td>
<td>Select the influxdb version to use.</td>
</tr>
<tr>
<td>Retention Policy</td>
<td>Select the retention policy. Default is "autogen" for v1.x and "infinite" for v2.x.</td>
</tr>
<tr>
<td>Hostname</td>
<td>The hostname to connect to the time-series server. Default is "localhost".</td>
</tr>
<tr>
<td>Port</td>
<td>The time-series database port. Default is 8086.</td>
</tr>
<tr>
<td>Database Name</td>
<td>The name of the database (v1.x) or bucket (v2.x) for Mycodo to store to and query from. Default is "mycodo_db".</td>
</tr>
<tr>
<td>Username</td>
<td>The username to access the database (if credentials are required). Default is "mycodo".</td>
</tr>
<tr>
<td>Password</td>
<td>The password to access the database (if credentials are required).</td>
</tr>
</tbody>
</table>

## Dashboard Settings

Page\: `[Gear Icon] -> Configure -> General`

<table>
<thead>
<tr class="header">
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Grid Cell Height (px)</td>
<td>The height of each widget cell, in pixels.</td>
</tr>
</tbody>
</table>

## Upgrade Settings

Page\: `[Gear Icon] -> Configure -> General`

<table>
<thead>
<tr class="header">
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Internet Test IP Address</td>
<td>The IP address to use to test for an active internet connection.</td>
</tr>
<tr>
<tr>
<td>Internet Test Port</td>
<td>The port to use to test for an active internet connection.</td>
</tr>
<tr>
<tr>
<td>Internet Test Timeout</td>
<td>The timeout period for testing for an active internet connection.</td>
</tr>
<tr>
<td>Check for Updates</td>
<td>Automatically check for updates every 2 days and notify through the web interface. If there is a new update, the Configure (Gear Icon) as well as the Upgrade menu will turn the color red.</td>
</tr>
</tbody>
</table>

## Energy Usage Settings

Page\: `[Gear Icon] -> Configure -> General`

In order to calculate accurate energy usage statistics, a few characteristics of your electrical system needs to be know. These variables should describe the characteristics of the electrical system being used by the relays to operate electrical devices.

!!! note
    If not using a current sensor, proper energy usage calculations will rely on the correct current draw to be set for each output (see [Output Settings](Outputs.md)).

<table>
<thead>
<tr class="header">
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Max Amps</td>
<td>Set the maximum allowed amperage to be switched on at any given time. If a output that's instructed to turn on will cause the sum of active devices to exceed this amount, the output will not be allowed to turn on, to prevent any damage that may result from exceeding current limits.</td>
</tr>
<tr>
<td>Voltage</td>
<td>Alternating current (AC) voltage that is switched by the outputs. This is usually 120 or 240.</td>
</tr>
<tr>
<td>Cost per kWh</td>
<td>This is how much you pay per kWh.</td>
</tr>
<tr>
<td>Currency Unit</td>
<td>This is the unit used for the currency that pays for electricity.</td>
</tr>
<tr>
<td>Day of Month</td>
<td>This is the day of the month (1-30) that the electricity meter is read (which will correspond to the electrical bill).</td>
</tr>
<tr>
<td>Generate Usage/Cost Report</td>
<td>These options define when an Energy Usage Report will be generated. Currently, these Only support the Output Duration calculation method.</td>
</tr>
<tr>
<td>Time Span to Generate</td>
<td>How often to automatically generate a usage/cost report.</td>
</tr>
<tr>
<td>Day of Week/Month to Generate</td>
<td>On which day of the week to generate the report. Daily: 1-7 (Monday = 1), Monthly: 1-28.</td>
</tr>
<tr>
<td>Hour of Day to Generate</td>
<td>On which hour of the day to generate the report (0-23).</td>
</tr>
</tbody>
</table>

## Controller Sample Rate Settings

Page\: `[Gear Icon] -> Configure -> General`

Each controller for Inputs, Outputs, and Functions operate periodically. The fastest these controllers can respond is determined by the sample rate of each. The looping function of each controller is paused for the specific duration. For instance, the Output controller can only have a resolution of 1 second if the sample rate is set to 1 second, meaning if you instruct an output to turn on or off, it will take a maximum of 1 second to respond to that request.

<table>
<thead>
<tr class="header">
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Max Amps</td>

CHANGELOG.md
## 8.15.6 (2023-01-30)

### Bugfixes

 - Revert SQLAlchemy to 1.4.46 due to 2.0.0 causing high CPU load


## 8.15.5 (2023-01-30)

### Bugfixes

 - Fix backend/frontend restart during Settings Import
 - Fix pumps not actuating with Regulate pH & Electrical Conductivity Function

### Features

 - Add listener and do_not_run_periodically as Function module options
 - Add Settings Import Log
 - Add ability to delete influxdb database on Diagnostics page

### Miscellaneous

 - Update Python packages
 - Update InfluxDB to 2.6.1
 - Switch from using libcamera-apps-lite to libcamera-apps ([#1270](https://github.com/kizniche/Mycodo/issues/1270))


## 8.15.4 (2023-01-19)

### Bugfixes

 - Fix potential issues with MycodoRelease Class
 - Fix library version typo in VL53L1X Input


## 8.15.3 (2023-01-18)

### Bugfixes

 - Fix regex to properly parse latest Mycodo release

### Miscellaneous

 - Update Python packages


## 8.15.2 (2023-01-18)

### Bugfixes

 - Fix PCF8574 Pump Output not properly determining channel states
 - Fix MQTT Subscribe Inputs resubscribing to MQTT topics after reconnect

### Features

 - Add feedback about why a Python package dependency was unmet when adding an Input/Output/Function

### Miscellaneous

 - Add library name to the title when multiple Inputs of the same name exist in the manual


## 8.15.1 (2023-01-14)

### Bugfixes

 - Fix endpoints of Widgets not able to be created until a frontend restart ([#1260](https://github.com/kizniche/Mycodo/issues/1260))


## 8.15.0 (2023-01-13)

This release fixes several bugs and changes how measurement timestamps are returned from the time-series database.

### InfluxDB Changes

Since support for InfluxDB 2.x was added, in order to make data that's returned from the time-series database more consistent, all timestamps from internal Mycodo functions will be returned as a decimal epoch value. Please update any custom modules or code to incorporate this change.

Previous behavior: InfluxDB 1.x returned data as a date/time string value (e.g. 2022-10-31T14:49:25.034447Z).

New Behavior: InfluxDB 1.x returned data as an epoch float value (e.g. 1667228575.557).

### Camera Changes

Cameras integration was initially built into the Mycodo system. This made it difficult for users to add new camera libraries or features unless they learned the codebase well enough to understand the many files needed to be modified. This release allows for camera functinality to be developed within a single Function module file. This enables users to more easily develop and implement their own Camera Functions or modify existing Camera Functions to add new features.

As this system is developed, the original camera system will be phased out, as it's a lot of work to maintain two different camera systems. As of now, there is only one Camera Function (libcamera). This should be considered EXPERIMENTAL and only used to test the development of this new feature until it is considered stable.

### Features

 - Add Input: MCP3008 (adafruit-circuitpython-mcp3xxx library)
 - Add Input: On/Off Output State
 - Add Input: Senseair K96 ([#1196](https://github.com/kizniche/Mycodo/issues/1196))
-  Add Input: VL53L4CD
 - Add Input: ams AS7341 light spectrum sensor
 - Add Output: Remote Mycodo Output: On/Off (API)
 - Add Output: Remote Mycodo Output: PWM (API)
 - Add Output: Sequent Microsystems 8-Relay HAT for Raspberry Pi

mycodo/config.py
# -*- coding: utf-8 -*-
#
# config.py - Global Mycodo settings
#
import binascii
import sys
from datetime import timedelta

import os
from flask_babel import lazy_gettext as lg

# Append proper path for other software reading this config file
sys.path.append(os.path.abspath(os.path.dirname(__file__)))
from config_translations import TRANSLATIONS as T

MYCODO_VERSION = '8.15.6'
ALEMBIC_VERSION = '16b28ef31b5b'

# FORCE UPGRADE MASTER
# Set True to enable upgrading to the master branch of the Mycodo repository.
# Set False to enable upgrading to the latest Release version (default).
# Do not use this feature unless you know what you're doing or have been
# instructed to do so, as it can really mess up your system.
FORCE_UPGRADE_MASTER = False

# Final release for each major version number
# Used to determine proper upgrade page to display
FINAL_RELEASES = ['5.7.3', '6.4.7', '7.10.0']

# ENABLE FLASK PROFILER
# Accessed at https://127.0.0.1/mycodo-flask-profiler
ENABLE_FLASK_PROFILER = False

# Install path (the parent directory of this file)
INSTALL_DIRECTORY = os.path.abspath(os.path.dirname(os.path.realpath(__file__)) + '/..')

# Database
DATABASE_NAME = "mycodo.db"
ALEMBIC_PATH = os.path.join(INSTALL_DIRECTORY, 'alembic_db')
DATABASE_PATH = os.path.join(INSTALL_DIRECTORY, 'databases')
ALEMBIC_UPGRADE_POST = os.path.join(ALEMBIC_PATH, 'alembic_post_upgrade_versions')
SQL_DATABASE_MYCODO = os.path.join(DATABASE_PATH, DATABASE_NAME)
MYCODO_DB_PATH = f'sqlite:///{SQL_DATABASE_MYCODO}'

# Misc paths
PATH_1WIRE = '/sys/bus/w1/devices/'
PATH_CONTROLLERS = os.path.join(INSTALL_DIRECTORY, 'mycodo/controllers')
PATH_FUNCTIONS = os.path.join(INSTALL_DIRECTORY, 'mycodo/functions')
PATH_ACTIONS = os.path.join(INSTALL_DIRECTORY, 'mycodo/actions')
PATH_INPUTS = os.path.join(INSTALL_DIRECTORY, 'mycodo/inputs')
PATH_OUTPUTS = os.path.join(INSTALL_DIRECTORY, 'mycodo/outputs')
PATH_WIDGETS = os.path.join(INSTALL_DIRECTORY, 'mycodo/widgets')
PATH_FUNCTIONS_CUSTOM = os.path.join(PATH_FUNCTIONS, 'custom_functions')
PATH_ACTIONS_CUSTOM = os.path.join(PATH_ACTIONS, 'custom_actions')
PATH_INPUTS_CUSTOM = os.path.join(PATH_INPUTS, 'custom_inputs')
PATH_OUTPUTS_CUSTOM = os.path.join(PATH_OUTPUTS, 'custom_outputs')
PATH_WIDGETS_CUSTOM = os.path.join(PATH_WIDGETS, 'custom_widgets')
PATH_USER_SCRIPTS = os.path.join(INSTALL_DIRECTORY, 'mycodo/user_scripts')
PATH_HTML_USER = os.path.join(INSTALL_DIRECTORY, 'mycodo/mycodo_flask/templates/user_templates')
PATH_PYTHON_CODE_USER = os.path.join(INSTALL_DIRECTORY, 'mycodo/user_python_code')
PATH_MEASUREMENTS_BACKUP = os.path.join(INSTALL_DIRECTORY, 'mycodo/backup_measurements')
PATH_SETTINGS_BACKUP = os.path.join(INSTALL_DIRECTORY, 'mycodo/backup_settings')
USAGE_REPORTS_PATH = os.path.join(INSTALL_DIRECTORY, 'output_usage_reports')
DEPENDENCY_INIT_FILE = os.path.join(INSTALL_DIRECTORY, '.dependency')
UPGRADE_INIT_FILE = os.path.join(INSTALL_DIRECTORY, '.upgrade')
BACKUP_PATH = '/var/Mycodo-backups'  # Where Mycodo backups are stored

# Log files
LOG_PATH = '/var/log/mycodo'  # Where generated logs are stored
LOGIN_LOG_FILE = os.path.join(LOG_PATH, 'login.log')
DAEMON_LOG_FILE = os.path.join(LOG_PATH, 'mycodo.log')
KEEPUP_LOG_FILE = os.path.join(LOG_PATH, 'mycodokeepup.log')
BACKUP_LOG_FILE = os.path.join(LOG_PATH, 'mycodobackup.log')
DEPENDENCY_LOG_FILE = os.path.join(LOG_PATH, 'mycododependency.log')
IMPORT_LOG_FILE = os.path.join(LOG_PATH, 'mycodoimport.log')
UPGRADE_LOG_FILE = os.path.join(LOG_PATH, 'mycodoupgrade.log')
UPGRADE_TMP_LOG_FILE = '/tmp/mycodoupgrade.log'
RESTORE_LOG_FILE = os.path.join(LOG_PATH, 'mycodorestore.log')
HTTP_ACCESS_LOG_FILE = '/var/log/nginx/access.log'
HTTP_ERROR_LOG_FILE = '/var/log/nginx/error.log'

# Lock files
LOCK_PATH = '/var/lock'
LOCK_FILE_STREAM = os.path.join(LOCK_PATH, 'mycodo-camera-stream.pid')

# Run files
RUN_PATH = '/var/run'
FRONTEND_PID_FILE = os.path.join(RUN_PATH, 'mycodoflask.pid')
DAEMON_PID_FILE = os.path.join(RUN_PATH, 'mycodo.pid')

# Remote admin
STORED_SSL_CERTIFICATE_PATH = os.path.join(
    INSTALL_DIRECTORY, 'mycodo/mycodo_flask/ssl_certs/remote_admin')

# Cameras
PATH_CAMERAS = os.path.join(INSTALL_DIRECTORY, 'cameras')

# Notes
PATH_NOTE_ATTACHMENTS = os.path.join(INSTALL_DIRECTORY, 'note_attachments')

# Determine if running in a Docker container
DOCKER_CONTAINER = os.environ.get('DOCKER_CONTAINER', False) == 'TRUE'

# Pyro5 URI/host, used by mycodo_client.py
if DOCKER_CONTAINER:
    PYRO_URI = 'PYRO:mycodo.pyro_server@mycodo_daemon:9080'
else:
    PYRO_URI = 'PYRO:mycodo.pyro_server@127.0.0.1:9080'

# Anonymous statistics
STATS_INTERVAL = 86400
STATS_HOST = 'fungi.kylegabriel.com'
STATS_PORT = 8086
STATS_USER = 'mycodo_stats'
STATS_PASSWORD = 'Io8Nasr5JJDdhPOj32222'
STATS_DATABASE = 'mycodo_stats'
STATS_CSV = os.path.join(INSTALL_DIRECTORY, 'statistics.csv')
ID_FILE = os.path.join(INSTALL_DIRECTORY, 'statistics.id')

# Login restrictions
LOGIN_ATTEMPTS = 5
LOGIN_BAN_SECONDS = 600  # 10 minutes

# Check for upgrade every 2 days (if enabled)
UPGRADE_CHECK_INTERVAL = 172800

TAGS_URL = 'https://api.github.com/repos/kizniche/Mycodo/git/refs/tags'

LANGUAGES = {
    'en': 'English',
    'de': 'Deutsche (German)',
    'es': 'Español (Spanish)',
    'fr': 'Français (French)',
    'id': 'Bahasa Indonesia (Indonesian)',
    'it': 'Italiano (Italian)',
    'nn': 'Norsk (Norwegian)',
    'nl': 'Nederlands (Dutch)',
    'pl': 'Polski (Polish)',
    'pt': 'Português (Portuguese)',
    'ru': 'русский язык (Russian)',
    'sr': 'српски (Serbian)',
    'sv': 'Svenska (Swedish)',
    'tr': 'Türkçe (Turkish)',
    'zh': '中文 (Chinese)'
}

DASHBOARD_WIDGETS = [
    ('', f"{lg('Add')} {lg('Dashboard')} {lg('Widget')}"),
    ('spacer', lg('Spacer')),
    ('graph', lg('Graph')),
    ('gauge', lg('Gauge')),
    ('indicator', T['indicator']['title']),
    ('measurement', T['measurement']['title']),
    ('output', T['output']['title']),
    ('output_pwm_slider', f"{T['output']['title']}: {lg('PWM Slider')}"),
    ('pid_control', lg('PID Control')),
    ('python_code', lg('Python Code')),
    ('camera', T['camera']['title'])
]

# Camera info
CAMERA_INFO = {
    'fswebcam': {
        'name': 'fswebcam',
        'dependencies_module': [
            ('apt', 'fswebcam', 'fswebcam')
        ],
        'capable_image': True,
        'capable_stream': False
    },
    'libcamera': {
        'name': 'libcamera',
        'dependencies_module': [
            ('apt', 'libcamera-apps', 'libcamera-apps')
        ],
        'capable_image': True,
        'capable_stream': False
    },
    'opencv': {
        'name': 'OpenCV',
        'dependencies_module': [
            ('pip-pypi', 'imutils', 'imutils==0.5.4'),
            ('apt', 'libgl1', 'libgl1'),
            ('pip-pypi', 'cv2', 'opencv-python==4.6.0.66')
        ],
        'capable_image': True,
        'capable_stream': True
    },
    'picamera': {
        'name': 'PiCamera (deprecated)',
        'dependencies_module': [
            ('pip-pypi', 'picamera', 'picamerab==1.13b1')
        ],
        'capable_image': True,
        'capable_stream': True
    },
    'raspistill': {
        'name': 'raspistill (deprecated)',
        'dependencies_module': [],
        'capable_image': True,
        'capable_stream': False
    },
    'http_address': {
        'name': 'URL (urllib)',
        'dependencies_module': [
            ('pip-pypi', 'imutils', 'imutils==0.5.4'),
            ('apt', 'libgl1', 'libgl1'),
            ('pip-pypi', 'cv2', 'opencv-python==4.6.0.66')
        ],
        'capable_image': True,
        'capable_stream': True
    },
    'http_address_requests': {
        'name': 'URL (requests)',
        'dependencies_module': [
            ('pip-pypi', 'imutils', 'imutils==0.5.4'),
            ('apt', 'libgl1', 'libgl1'),
            ('pip-pypi', 'cv2', 'opencv-python==4.6.0.66')
        ],
        'capable_image': True,
        'capable_stream': False
    },
}

METHOD_DEP_BASE = [
    ('bash-commands',
     ['/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highcharts-9.1.2.js'],
     [
        'wget --no-clobber https://code.highcharts.com/zips/Highcharts-9.1.2.zip',
        'unzip Highcharts-9.1.2.zip -d Highcharts-9.1.2',
        'cp -rf Highcharts-9.1.2/code/highcharts.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highcharts-9.1.2.js',
        'cp -rf Highcharts-9.1.2/code/highcharts.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highcharts.js.map',
        'rm -rf Highcharts-9.1.2'
     ])
]

# Method info
METHOD_INFO = {
    'Date': {
        'name': lg('Time/Date'),
        'dependencies_module': METHOD_DEP_BASE
    },
    'Duration': {
        'name': lg('Duration'),
        'dependencies_module': METHOD_DEP_BASE
    },
    'Daily': {
        'name': f"{lg('Daily')} ({lg('Time-Based')})",
        'dependencies_module': METHOD_DEP_BASE
    },
    'DailySine': {
        'name': f"{lg('Daily')} ({lg('Sine Wave')})",
        'dependencies_module': METHOD_DEP_BASE
    },
    'DailyBezier': {
        'name': f"{lg('Daily')} ({lg('Bezier Curve')})",
        'dependencies_module': [
            ('apt', 'libatlas-base-dev', 'libatlas-base-dev'),
            ('pip-pypi', 'numpy', 'numpy==1.22.3')
        ] + METHOD_DEP_BASE
    },
    'Cascade': {
        'name': lg('Method Cascade'),
        'dependencies_module': METHOD_DEP_BASE
    }
}

# Method form dropdown
METHODS = [
    ('Date', METHOD_INFO['Date']['name']),
    ('Duration', METHOD_INFO['Duration']['name']),
    ('Daily', METHOD_INFO['Daily']['name']),
    ('DailySine', METHOD_INFO['DailySine']['name']),
    ('DailyBezier', METHOD_INFO['DailyBezier']['name']),
    ('Cascade', METHOD_INFO['Cascade']['name'])
]

PID_INFO = {
    'measure': {
        0: {
            'measurement': '',
            'unit': '',
            'name': f"{T['setpoint']['title']}",
            'measurement_type': 'setpoint'
        },
        1: {
            'measurement': '',
            'unit': '',
            'name': f"{T['setpoint']['title']} ({lg('Band Min')})",
            'measurement_type': 'setpoint'
        },
        2: {
            'measurement': '',
            'unit': '',
            'name': f"{T['setpoint']['title']} ({lg('Band Max')})",
            'measurement_type': 'setpoint'
        },
        3: {
            'measurement': 'pid_p_value',
            'unit': 'pid_value',
            'name': 'P-value'
        },
        4: {
            'measurement': 'pid_i_value',
            'unit': 'pid_value',
            'name': 'I-value'
        },
        5: {
            'measurement': 'pid_d_value',
            'unit': 'pid_value',
            'name': 'D-value'
        },
        6: {
            'measurement': 'duration_time',
            'unit': 's',
            'name': f"{T['output']['title']} ({T['duration']['title']})"
        },
        7: {
            'measurement': 'duty_cycle',
            'unit': 'percent',
            'name': f"{T['output']['title']} ({T['duty_cycle']['title']})"
        },
        8: {
            'measurement': 'volume',
            'unit': 'ml',
            'name': f"{T['output']['title']} ({T['volume']['title']})"
        },
        9: {
            'measurement': 'unitless',
            'unit': 'none',
            'name': f"{T['output']['title']} ({T['value']['title']})"
        }
    }
}

DEPENDENCIES_GENERAL = {
    'highstock': {
        'name': 'Highstock',
        'dependencies_module': [
            ('bash-commands',
             [
                 '/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highstock-9.1.2.js',
                 '/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highcharts-more-9.1.2.js',
                 '/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/data-9.1.2.js',
                 '/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/exporting-9.1.2.js',
                 '/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/export-data-9.1.2.js',
                 '/var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/offline-exporting-9.1.2.js'
             ],
             [
                 'wget --no-clobber https://code.highcharts.com/zips/Highcharts-Stock-9.1.2.zip',
                 'unzip Highcharts-Stock-9.1.2.zip -d Highcharts-Stock-9.1.2',
                 'cp -rf Highcharts-Stock-9.1.2/code/highstock.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highstock-9.1.2.js',
                 'cp -rf Highcharts-Stock-9.1.2/code/highstock.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highstock.js.map',
                 'cp -rf Highcharts-Stock-9.1.2/code/highcharts-more.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highcharts-more-9.1.2.js',
                 'cp -rf Highcharts-Stock-9.1.2/code/highcharts-more.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/highcharts-more.js.map',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/data.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/data-9.1.2.js',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/data.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/data.js.map',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/exporting.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/exporting-9.1.2.js',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/exporting.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/exporting.js.map',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/export-data.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/export-data-9.1.2.js',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/export-data.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/export-data.js.map',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/offline-exporting.js /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/offline-exporting-9.1.2.js',
                 'cp -rf Highcharts-Stock-9.1.2/code/modules/offline-exporting.js.map /var/mycodo-root/mycodo/mycodo_flask/static/js/user_js/offline-exporting.js.map',
                 'rm -rf Highcharts-Stock-9.1.2'
             ])
        ]
    }
}

# Conditional Functions
CONDITIONAL_CONDITIONS = [
    ('measurement',
     f"{T['measurement']['title']} ({T['single']['title']}, {T['last']['title']})"),
    ('measurement_past_average',
     f"{T['measurement']['title']} ({T['single']['title']}, {T['past']['title']}, {T['average']['title']})"),
    ('measurement_past_sum',
     f"{T['measurement']['title']} ({T['single']['title']}, {T['past']['title']}, {T['sum']['title']})"),
    ('measurement_dict',
     f"{T['measurement']['title']} ({T['multiple']['title']}, {T['past']['title']})"),
    ('gpio_state', lg('GPIO State')),
    ('output_state', lg('Output State')),
    ('output_duration_on', lg('Output Duration On')),
    ('controller_status', lg("Controller Running")),
]

FUNCTION_INFO = {
    'function_actions': {
        'name': lg('Execute Actions'),
        'dependencies_module': []
    },
    'conditional_conditional': {
        'name': f"{T['conditional']['title']} {T['controller']['title']}",
        'dependencies_module': [
            ('pip-pypi', 'pylint', 'pylint==2.12.2')
        ]
    },
    'pid_pid': {
        'name': f"{T['pid']['title']} {T['controller']['title']}",
        'dependencies_module': []
    },
    'trigger_edge': {
        'name': f"{T['trigger']['title']}: {T['edge']['title']}",
        'dependencies_module': []
    },
    'trigger_output': {
        'name': f"{T['trigger']['title']}: {T['output']['title']} ({T['on']['title']}/{T['off']['title']})",
        'dependencies_module': []
    },
    'trigger_output_pwm': {
        'name': f"{T['trigger']['title']}: {T['output']['title']} ({T['pwm']['title']})",
        'dependencies_module': []
    },
    'trigger_timer_daily_time_point': {
        'name': f"{T['trigger']['title']}: {lg('Daily Time Point')}",
        'message': 'Time Point Trigger Functions will execute Actions every day at the specified Time.',
        'dependencies_module': []
    },
    'trigger_timer_daily_time_span': {
        'name': f"{T['trigger']['title']}: {lg('Daily Time Span')}",
        'message': 'Time Span Trigger Functions will execute Actions every day within the set Start and End times, every Period. The first execution will be the Start Time.',
        'dependencies_module': []
    },
    'trigger_timer_duration': {
        'name': f"{T['trigger']['title']}: {T['duration']['title']}",
        'message': 'Duration Trigger Functions will execute Actions every Period from when it is activated. A start offset can be added to delay the first execution.',
        'dependencies_module': []
    },
    'trigger_run_pwm_method': {
        'name': f"{T['trigger']['title']}: {lg('Run PWM Method')}",
        'dependencies_module': []
    },
    'trigger_sunrise_sunset': {
        'name': f"{T['trigger']['title']}: {lg('Sunrise/Sunset')}",
        'dependencies_module': [
            ('pip-pypi', 'suntime', 'suntime==1.2.5')
        ]
    }
}

FUNCTIONS = [
    ('function_actions', FUNCTION_INFO['function_actions']['name']),
    ('conditional_conditional', FUNCTION_INFO['conditional_conditional']['name']),
    ('pid_pid', FUNCTION_INFO['pid_pid']['name']),
    ('trigger_edge', FUNCTION_INFO['trigger_edge']['name']),
    ('trigger_output', FUNCTION_INFO['trigger_output']['name']),
    ('trigger_output_pwm', FUNCTION_INFO['trigger_output_pwm']['name']),
    ('trigger_timer_daily_time_point', FUNCTION_INFO['trigger_timer_daily_time_point']['name']),
    ('trigger_timer_daily_time_span', FUNCTION_INFO['trigger_timer_daily_time_span']['name']),
    ('trigger_timer_duration', FUNCTION_INFO['trigger_timer_duration']['name']),
    ('trigger_run_pwm_method', FUNCTION_INFO['trigger_run_pwm_method']['name']),
    ('trigger_sunrise_sunset', FUNCTION_INFO['trigger_sunrise_sunset']['name'])
]

# User Roles
USER_ROLES = [
    dict(id=1, name='Admin',
         edit_settings=True, edit_controllers=True, edit_users=True,
         view_settings=True, view_camera=True, view_stats=True, view_logs=True,
         reset_password=True),
    dict(id=2, name='Editor',
         edit_settings=True, edit_controllers=True, edit_users=False,
         view_settings=True, view_camera=True, view_stats=True, view_logs=True,
         reset_password=True),
    dict(id=3, name='Monitor',
         edit_settings=False, edit_controllers=False, edit_users=False,
         view_settings=True, view_camera=True, view_stats=True, view_logs=True,
         reset_password=True),
    dict(id=4, name='Guest',
         edit_settings=False, edit_controllers=False, edit_users=False,
         view_settings=False, view_camera=False, view_stats=False, view_logs=False,
         reset_password=False),
    dict(id=5, name='Kiosk',
         edit_settings=False, edit_controllers=False, edit_users=False,
         view_settings=False, view_camera=True, view_stats=True, view_logs=False,
         reset_password=False)
]

# Web UI themes
THEMES = [
    ('cerulean', 'Cerulean'),
    ('cosmo', 'Cosmo'),
    ('cyborg', 'Cyborg'),
    ('darkly', 'Darkly'),
    ('flatly', 'Flatly'),
    ('journal', 'Journal'),
    ('literia', 'Literia'),
    ('lumen', 'Lumen'),
    ('lux', 'Lux'),
    ('materia', 'Materia'),
    ('minty', 'Minty'),
    ('pulse', 'Pulse'),
    ('sandstone', 'Sandstone'),
    ('simplex', 'Simplex'),
    ('slate', 'Slate'),
    ('solar', 'Solar'),
    ('spacelab', 'Spacelab'),
    ('superhero', 'Superhero'),
    ('united', 'United'),
    ('yeti', 'Yeti')
]

THEMES_DARK = ['cyborg', 'darkly', 'slate', 'solar', 'superhero']


class ProdConfig(object):
    """Production Configuration."""
    SQL_DATABASE_MYCODO = os.path.join(DATABASE_PATH, DATABASE_NAME)
    MYCODO_DB_PATH = f'sqlite:///{SQL_DATABASE_MYCODO}'
    SQLALCHEMY_DATABASE_URI = f'sqlite:///{SQL_DATABASE_MYCODO}'
    SQLALCHEMY_TRACK_MODIFICATIONS = False

    FLASK_PROFILER = {
        "enabled": True,
        "storage": {
            "engine": "sqlalchemy",
            "db_url": f"sqlite:///{os.path.join(DATABASE_PATH, 'profile.db')}"
        },
        "basicAuth": {
            "enabled": True,
            "username": "admin231",
            "password": "admin421378956"
        },
        "ignore": [
            "^/static/.*",
            "/login",
            "/settings/users"
        ],
        "endpointRoot": "mycodo-flask-profiler"
    }

    WTF_CSRF_TIME_LIMIT = 60 * 60 * 24 * 7  # 1 week expiration
    REMEMBER_COOKIE_DURATION = timedelta(days=90)
    SESSION_TYPE = "filesystem"

    # Ensure file containing the Flask secret_key exists
    FLASK_SECRET_KEY_PATH = os.path.join(DATABASE_PATH, 'flask_secret_key')
    if not os.path.isfile(FLASK_SECRET_KEY_PATH):
        secret_key = binascii.hexlify(os.urandom(32)).decode()
        if not os.path.exists(DATABASE_PATH):
            os.makedirs(DATABASE_PATH)
        with open(FLASK_SECRET_KEY_PATH, 'w') as file:
            file.write(secret_key)
    SECRET_KEY = open(FLASK_SECRET_KEY_PATH, 'rb').read()


class TestConfig(object):
    """Testing Configuration."""
    SQLALCHEMY_DATABASE_URI = 'sqlite://'  # in-memory db only. tests drop the tables after they run
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    PRESERVE_CONTEXT_ON_EXCEPTION = False
    RATELIMIT_ENABLED = False
    SECRET_KEY = '1234'
    SESSION_TYPE = "filesystem"
    TESTING = True
    DEBUG = True
